### ISSUE TITLE
New Yank class interface

### ISSUE THREAD

**Issue by jchodera** [choderalab/yank#102]
For #94, I am reworking the `Yank` class interface.

Processing associated with the new command-line interface has all been moved into a `yank/commands/` submodule, but this leaves open the question of how one might want to use Yank through a Python interface.  Since we don't have any applications that do this right now, this division is currently worryingly arbitrary.

Here's a sketch of how the current interface is used:
```
# Create a Yank object, binding it to a directory where all simulation data will be stored.
yank = Yank(store_directory)
# Create a new simulation.
phases = ['solvent', 'complex'] # thermodynamic legs to simulate
systems = { 'solvent' : solvent_system, 'complex' : complex_system } # fully interacting systems
positions = { 'solvent' : solvent_positons, 'complex' : complex_positions } # positions could be lists of positions to allocate to replicas
atom_indices = { 'solvent' : { 'ligand' : ... }, 'complex' : {'ligand' : ..., 'receptor' : ...} } # atom_indices[phase][component] gives atom indices by phase and system component
yank.create(phases, systems, positions, alchemical_indices, thermodynamic_state)
# Run a few iterations of a given phase.
yank.run(phase='complex', niterations=10)
# Get status / progress info.
status = yank.status()
# Run all phases to completion..
yank.run(mpicomm) # Only one 'run' now, optionally takes 'mpicomm'
# Programmatically retrieve results.
results = yank.analyze()

# Resume an existing simulation, to resume or extend.
yank = Yank(store_directory)
options = { 'niterations' : 10000 }
yank.resume(options)
yank.run()
```
But is this how we would actually want to use it?

Also, what about things like hydration free energies?  I'm trying to move `Yank` toward a more general notion of just managing replica-exchange simulations of multiple thermodynamic legs (we currently call them "phases"), but there would be calculation-dependent logic that would be encoded by the *name* of the phase, which could either be useful or weird.

For hydration free energy calculations, we would have
```
phases = ['vacuum', 'solvent']
```
while a binding free energy calculation would have
```
phases = ['solvent', 'complex']
```
where either the phase name `'complex'` or the presence of both `receptor` and `ligand` components in the `atom_indices` and lack of explicit solvent would trigger an addition of protein-ligand restraints.
We could potentially mix these together too to get both hydration and binding free energies.

Receptor selectivities obtained with
```
phases = ['complex1', 'complex2']
```
etc.

In summary, the question is "how do we make the Python interface of `Yank` actually useful?

**Comment by jchodera** [comment]
Handled by #94
